src/tunnel.cc

```diff
@@ -361,6 +393,61 @@ TunnelStateData::Connection::~Connection()
     safe_free(buf);
 }
 
+bool
+TunnelStateData::checkRetry()
+{
+    if (shutting_down)
+        return false;
+    if (!retriable)
+        return false;
+    if (noConnections())
+        return false;
+    return true;
+}
+
+void
+TunnelStateData::retryOrBail(const char *context)
+{
+    server.closeQuietly(); // may already be closed
+
+    if (checkRetry()) {
+        if (!destinations->empty())
+            return startConnecting(); // try connecting to another destination
+
+        if (subscribed) {
+            debugs(26, 4, "wait for more destinations to try");
+            return; // expect a noteDestination*() call
+        }
+
+        // fall through to bail
+    }
+
+    /* bail */
+
+    if (request)
+        request->hier.stopPeerClock(false);
+
+    // TODO: Add sendSavedErrorOr(err_type type, Http::StatusCode, context).
+    // Then, the remaining method code (below) should become the common part of
+    // sendNewError() and sendSavedErrorOr(), used in "error detected" cases.
+    const auto error = savedError ? savedError : new ErrorState(ERR_CANNOT_FORWARD,
```

Comment:
It looks like the created error is never deleted if canSendError is false.

Reply:
Agreed. Please fix. I suspect calling saveError(error) if savedError is nil would be sufficient, but I did not check the details. The TODO above this line is still out of scope.

Reply:
Fixed at d38e99c.